app/scripts/modules/core/src/pipeline/config/triggers/concourse/ConcourseTrigger.tsx

```diff
@@ -1,205 +1,134 @@
-    }
-  };
-
-  private fetchAvailablePipelines = (trigger: IConcourseTrigger) => {
-    const { master, team } = trigger;
-    if (master && team) {
-      Observable.fromPromise(ConcourseService.listPipelinesForTeam(master, team))
-        .takeUntil(this.destroy$)
-        .subscribe(pipelines => this.setState({ pipelines }));
-    }
-  };
-
-  private fetchAvailableJobs = (trigger: IConcourseTrigger) => {
-    const { master, project, team } = trigger;
-    if (master && team && project) {
-      const pipeline = project.split('/').pop();
-      Observable.fromPromise(ConcourseService.listJobsForPipeline(master, team, pipeline))
-        .takeUntil(this.destroy$)
-        .subscribe(jobs => this.setState({ jobs }));
-    }
-  };
-
-  public render() {
-    const { jobName, team, project, master } = this.props.trigger;
-    const { jobs, pipelines, teams, masters } = this.state;
-    const pipeline = project && project.split('/').pop();
-
-    const jobOptions = jobs.map((name: string) => {
-      const jobNameOption = name.split('/').pop();
-      return { label: jobNameOption, value: jobNameOption };
-    });
-
-    return (
-      <>
-        <FormField
-          label="Master"
-          value={master}
-          onChange={e => this.onMasterChanged(e.target.value)}
-          input={props => (
-            <ReactSelectInput {...props} placeholder="Select a master..." stringOptions={masters} clearable={false} />
-          )}
-        />
-
-        <FormField
-          label="Team"
-          value={team}
-          onChange={e => this.onTeamChanged(e.target.value)}
-          input={props =>
-            master ? (
-              <ReactSelectInput {...props} placeholder="Select a team..." stringOptions={teams} clearable={false} />
-            ) : (
-              <p className="form-control-static">(Select a master)</p>
-            )
-          }
-        />
-
-        <FormField
-          label="Pipeline"
-          value={pipeline}
-          onChange={e => this.onPipelineChanged(e.target.value)}
-          input={props =>
-            team ? (
-              <ReactSelectInput
-                {...props}
-                placeholder="Select a pipeline..."
-                stringOptions={pipelines}
-                clearable={false}
-              />
-            ) : (
-              <p className="form-control-static">(Select a master and team)</p>
-            )
-          }
-        />
-
-        <FormField
-          label="Job"
-          value={jobName}
-          onChange={e => this.onJobChanged(e.target.value)}
-          input={props =>
-            pipeline ? (
-              <ReactSelectInput {...props} placeholder="Select a job..." options={jobOptions} clearable={false} />
-            ) : (
-              <p className="form-control-static">(Select a master, team, and pipeline)</p>
-            )
-          }
-        />
-      </>
-    );
+  // Fetches data when anything in the dependencies list changes
```

Comment:
Ah yes, I was experimenting here.  I noticed a pattern where I wanted to make an async call, but not until all the inputs were available.  This `useData` hook builds on the `useLatestPromise` hook (which itself builds on `useEffect`) by 1) waiting to make the request until all deps exist, and 2) returning the default value otherwise.

This hook makes some assumptions and provides opinions.  In some other spots in this PR I did use a similar pattern, but wrote it out longhand, such as:
```js
 const fetchJobs = useLatestPromise(() => {
    return master ? IgorService.listJobsForMaster(master) : null;
  }, [master]);
```
or
```js
  const fetchPipelines = useLatestPromise(() => {
    return application ? PipelineConfigService.getPipelinesForApplication(application) : null;
  }, [application]);
```
or even this one which was written a bit differently:
```js
  const fetchJobs = useLatestPromise(() => {
    return master && IgorService.listJobsForMaster(master).then(x => x.map(getJobParts));
  }, [master]);
```

Seems like it might be useful to pull out and remove.  It does remove some boilerplate, but also forces some opinions. WDYT

Reply:
Got it! I'll keep this in the back of my mind moving forward and if we start to see a bunch of places where we're writing out that logic longhand might be worth pulling out and explicitly documenting with your comments above. Thanks as always for the super helpful explanation.